File: native/android/app/src/main/java/br/com/azulzin/app/PushComponent.kt
package br.com.azulzin.app

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat
import com.google.firebase.FirebaseApp
import com.google.firebase.messaging.FirebaseMessaging
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination

// The "push" bridge component (.plans/mobile/04 §3): the web's Avisos row sends
// "register" (may prompt POST_NOTIFICATIONS); the native layout sends
// "registerIfGranted" on launch (silent). Replies carry the FCM token; the WEB posts it
// to /push_devices through its own session — no native HTTP, no cookie extraction.
//
// ponytail: Firebase initializes only once the founder ships google-services.json
// (AzulzinApplication guards init). Until then no token reply is ever sent.
class PushComponent(name: String, private val delegate: BridgeDelegate<HotwireDestination>) :
    BridgeComponent<HotwireDestination>(name, delegate) {

    override fun onReceive(message: Message) {
        when (message.event) {
            "register" -> register(message, interactive = true)
            "registerIfGranted" -> register(message, interactive = false)
        }
    }

    private fun register(message: Message, interactive: Boolean) {
        val activity = delegate.destination.fragment.activity as? MainActivity ?: return
        if (FirebaseApp.getApps(activity).isEmpty()) return
        val needsPermission = Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS) !=
            PackageManager.PERMISSION_GRANTED
        if (needsPermission) {
            if (interactive) activity.requestNotificationsPermission { granted ->
                if (granted) replyWithToken(message)
            }
        } else {
            replyWithToken(message)
        }
    }

    private fun replyWithToken(message: Message) {
        FirebaseMessaging.getInstance().token.addOnSuccessListener { token ->
            replyTo(message.event,
                """{"token":"$token","platform":"android","appVersion":"${BuildConfig.VERSION_NAME}"}""")
        }
    }
}

File: native/ios/app/app/PushComponent.swift
import HotwireNative
import UIKit
import UserNotifications
#if canImport(FirebaseMessaging)
import FirebaseMessaging
#endif

// The "push" bridge component (.plans/mobile/04 §3): the web's Avisos row sends
// "register" (may prompt for OS permission); the native layout sends "registerIfGranted"
// on launch (silent — token rotation must never strand a device). Replies carry the FCM
// token; the WEB posts it to /push_devices through its own session.
//
// ponytail: FirebaseMessaging is compiled in only once the founder adds the SPM package
// + GoogleService-Info.plist (a Firebase project is founder-provisioned). Until then the
// permission flow works but no token reply is sent — the web posts nothing.
final class PushComponent: BridgeComponent {
    override nonisolated class var name: String { "push" }

    override func onReceive(message: Message) {
        switch message.event {
        case "register":          register(message: message, interactive: true)
        case "registerIfGranted": register(message: message, interactive: false)
        default: break
        }
    }

    private func register(message: Message, interactive: Bool) {
        let center = UNUserNotificationCenter.current()
        center.getNotificationSettings { settings in
            switch settings.authorizationStatus {
            case .notDetermined:
                guard interactive else { return }
                center.requestAuthorization(options: [.alert, .badge, .sound]) { granted, _ in
                    guard granted else { return }
                    self.replyWithToken(message)
                }
            case .authorized, .provisional, .ephemeral:
                self.replyWithToken(message)
            default:
                break   // denied — the OS Settings app is the only way back
            }
        }
    }

    private func replyWithToken(_ message: Message) {
        DispatchQueue.main.async { UIApplication.shared.registerForRemoteNotifications() }
        #if canImport(FirebaseMessaging)
        Messaging.messaging().token { token, _ in
            guard let token else { return }
            let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0"
            let data = #"{"token":"\#(token)","platform":"ios","appVersion":"\#(version)"}"#
            self.reply(to: message.event, with: data)
        }
        #endif
    }
}
